src/check-dsl.ts

```diff
@@ -52,6 +54,327 @@ const defaultError = (lines: string[]) => {
+        const [decodedType, decodedRelation] = destructTupleToUserset(item);
+        if (decodedRelation) {
+          if (!transformedTypes[decodedType] || !transformedTypes[decodedType].relations[decodedRelation]) {
+            // type/relation is not defined
+            const typeIndex = getTypeLineNumber(type, lines);
+            const lineIndex = getRelationLineNumber(relation, lines, typeIndex);
+            reporter.invalidTypeRelation({
+              lineIndex,
+              value: `${decodedType}#${decodedRelation}`,
+              typeName: decodedType,
+              relationName: decodedRelation,
+            });
+          }
+        } else if (!transformedTypes[decodedType]) {
+          // type is not defined
+          const typeIndex = getTypeLineNumber(type, lines);
+          const lineIndex = getRelationLineNumber(relation, lines, typeIndex);
+          reporter.invalidType({
+            lineIndex,
+            value: `${decodedType}`,
+            typeName: decodedType,
+          });
+        }
+      }
+      break;
+    }
+    case RewriteType.ComputedUserset: {
+      if (childDef.target && !transformedTypes[type].relations[childDef.target]) {
+        const typeIndex = getTypeLineNumber(type, lines);
+        const lineIndex = getRelationLineNumber(relation, lines, typeIndex);
+        const value = childDef.target;
+        reporter.invalidRelation({
+          lineIndex,
+          value,
+          validRelations: Object.keys(transformedTypes[type].relations),
+        });
+      }
+      break;
+    }
+    case RewriteType.TupleToUserset: {
+      // for this case, we need to consider both the "from" and "relation"
+      if (childDef.from && childDef.target) {
+        // First, check to see if the childDef.from exists
+        if (!transformedTypes[type].relations[childDef.from]) {
+          const typeIndex = getTypeLineNumber(type, lines);
+          const lineIndex = getRelationLineNumber(relation, lines, typeIndex);
+          reporter.invalidTypeRelation({
+            lineIndex,
+            value: `${childDef.target} from ${childDef.from}`,
+            typeName: type,
+            relationName: childDef.from,
+          });
+        } else {
+          const [fromTypes, isValid] = allowableTypes(transformedTypes, type, childDef.from);
+          if (isValid) {
+            for (const item of fromTypes) {
+              const [decodedType, decodedRelation] = destructTupleToUserset(item);
+              if (decodedRelation) {
+                const typeIndex = getTypeLineNumber(type, lines);
+                const lineIndex = getRelationLineNumber(relation, lines, typeIndex);
+                reporter.tupleUsersetRequiresDirect({
+                  lineIndex,
+                  value: childDef.from,
+                });
+              } else {
+                if (!transformedTypes[decodedType] || !transformedTypes[decodedType].relations[childDef.target]) {
```

Comment:
Not sure if we need this here.

The validation for that definition should handle if the type is there or not. This should just validate that there are not usersets allowed

Reply:
We do...let's say we have
```
type A
type B
  relations
    define R1: [A] as self
type C
  relations
    define R2: [A] as self
    define R3 as R1 from R2
```

In this case, the type is present but A cannot have R1.

Reply:
Gotcha 👍🏽